### PR TITLE
fix: filtering promise properties in object literal completion

### LINKED ISSUE #59298: Avoid providing promise methods as completions at return positions in non-async functions
### 🔎 Search Terms

completions promise return expression type async members

### 🕗 Version & Regression Information

- This is the behavior in every version I tried

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=5.6.0-dev.20240716#code/C4TwDgpgBAsiAq5oF4oG8CwAoKUBmA9gQFxQDOwATgJYB2A5gNzYC+zW2eArrQMbDUCtKMAgUAFLwBGpcQEooyAHywESKAB8oABUoEAttTIQAPHESQlCtC2zZRE+YpXjMOKAHoAVAEYvH1jk5diA

### 💻 Code

```ts
type MyType = {
  foo: string;
};

function test(cb: () => MyType | Promise<MyType>) {}

test(() => ({
  /*1*/
}));
```

### 🙁 Actual behavior

Completions provided at the marker position include `catch`, `foo` and `then`

### 🙂 Expected behavior

It would be great if in a situation like this promise methods would get at least deprioritized and if they would appear at the bottom of the list.

### Additional information about the issue

It's technically okay to return thenables compatible with `Promise` interface but it should be a very rare situation. I think a DX win here would justify hiding those members completely.

### PR DESCRIPTION
**Pull request by syi0808**
### screenshot
스크린샷 2024-07-17 오후 4 08 23

~~In the actual development environment, it was possible to reproduce what was solved.~~
~~But I can't find a case to write a test. Can someone give me an idea?~~

Fixes #59298

### CODE REVIEW

**Comment by Andarist** [comment]
Tests for completions can be written like this: 
https://github.com/microsoft/TypeScript/blob/f37482cd16118409284e015409deaf14b416b4a0/tests/cases/fourslash/completionsPropertiesPriorities.ts

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -31176,6 +31176,14 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     return contextualAwaitedType && getUnionType([contextualAwaitedType, createPromiseLikeType(contextualAwaitedType)]);
                 }
 
+                const symbol = getSymbolForExpression(node);
+
+                if (symbol && (symbol.flags & SymbolFlags.ObjectLiteral)) {
+                    return filterType(contextualReturnType, t => {
+                        return !getAwaitedTypeOfPromise(t);
+                    });
```
Comment (Andarist) [inline src/compiler/checker.ts:31184]: Personally, I don't think this is the right fix. TS is a structural type at heart so this doesn't seem to be right - even if it's not overly practical for such types to contribute to contextual typing here... there is no type system rule that forbids it.

Completions are a separate thing and that could potentially be handled elsewhere (https://github.com/microsoft/TypeScript/blob/f37482cd16118409284e015409deaf14b416b4a0/src/services/completions.ts)

Comment (syi0808) [reply src/compiler/checker.ts:31184]: Thank you very much for taking the time to review it. Oh There was a scope called "completion", i didn't know sorry.
I think this is related to autocomplete, so I think it's right to revise that part. I'll turn this PR to draft and research it more.

Comment (syi0808) [reply src/compiler/checker.ts]: I modified this inside the completion.

I'd appreciate it if you could give me your opinion when you have time!

File: src/compiler/types.ts [hunk 2]
```diff
@@ -4973,6 +4973,7 @@ export interface TypeChecker {
     getPromisedTypeOfPromise(promise: Type, errorNode?: Node): Type | undefined;
     /** @internal */
     getAwaitedType(type: Type): Type | undefined;
+    getAwaitedTypeOfPromise(type: Type): Type | undefined;
```
Comment (jakebailey) [inline src/compiler/types.ts:4976]: This should be internal if it's actually needed.

File: src/compiler/types.ts [hunk 3]
```diff
@@ -5288,6 +5289,8 @@ export interface TypeChecker {
     tryGetThisTypeAt(node: Node, includeGlobalThis?: boolean, container?: ThisContainer): Type | undefined;
     /** @internal */ getTypeArgumentConstraint(node: TypeNode): Type | undefined;
 
+    filterType(type: Type, f: (t: Type) => boolean): Type;
```
Comment (jakebailey) [inline src/compiler/types.ts:5292]: Eek, definitely not sure about putting this in the exported interface, let alone the public API...

Comment (syi0808) [reply src/compiler/types.ts:5292]: Okay, I will find a way to solve the completion internally without creating more public api.

**Comment by jakebailey** [comment]
This looks good to me, though #59316 was accepted without determining whether or not the items should be prioritized or filtered, and this PR filters them out entirely.

I assume that ordering them is going to be a challenge, so are we okay with not providing promise-like method completions here?

**Comment by andrewbranch** [comment]
@typescript-bot pack this

File: src/services/completions.ts [hunk 4]
```diff
@@ -5472,9 +5472,10 @@ function getJsDocTagAtPosition(node: Node, position: number): JSDocTag | undefin
 /** @internal */
 export function getPropertiesForObjectExpression(contextualType: Type, completionsType: Type | undefined, obj: ObjectLiteralExpression | JsxAttributes, checker: TypeChecker): Symbol[] {
     const hasCompletionsType = completionsType && completionsType !== contextualType;
+    const promiseFilteredContextualType = checker.getUnionType(filter(contextualType.flags & TypeFlags.Union ? (contextualType as UnionType).types : [contextualType], t => !checker.getAwaitedType(t) || !checker.getPromisedTypeOfPromise(t)));
```
Comment (andrewbranch) [inline src/services/completions.ts:5475]: Why are both `getAwaitedType` and `getPromisedTypeOfPromise` needed?

Comment (syi0808) [reply src/services/completions.ts]: Looking at it again, I don't think I need the `getAwaitedType`. I deleted it. Thank you.

**Review by andrewbranch — APPROVED** [review]

**Review by jakebailey — APPROVED** [review]

**Pull request merged**